[Felix Ong] iP

felix-ong opened the pull request:
# Duke
> "The best task management tool."

Duke is a tool designed to manage your tasks. It's:
* text-based
* fast
* easy to use
* **FREE**

To start using Duke,
1. Download it from [here](https://github.com/felix-ong/ip)
2. Run the jar file
3. Start typing in commands! 😄 

Features:
- [x] Managing todos, events and deadlines
- [ ] Reminders

Duke can also be used for learning Java. Here's the `main` method.
```java
public static void main(String[] args) {
    new Duke("data/tasks.txt").run();
}
```

Commit:
Add Gradle support

felix-ong committed:
Implement skeletal version of Duke

felix-ong committed:
Add the ability to store whatever text entered by the user and
display them back to the user when requested

felix-ong committed:
Add the ability to mark tasks as done
Add Task class to represent tasks

felix-ong committed:
Add support for tracking three types of tasks: Todo, Deadline, Event

felix-ong committed:
Add semi-automated text UI testing

felix-ong committed:
Add error handling for two types of errors
Add DukeException class, handle unknown command
and todo without description

felix-ong committed:
Add support for deleting tasks from the list

felix-ong committed:
Fix excessive line length of code

felix-ong committed:
Add load data functionality

felix-ong committed:
Add ability to save the tasks in the hard disk automatically whenever the task list changes

felix-ong committed:
Change delete tasks to save tasks automatically

felix-ong committed:
Change deadline and event dates from String to java.time.LocalDate

felix-ong committed:
Merge branch 'branch-Level-7'

felix-ong committed:
Merge branch 'branch-Level-8'

felix-ong committed:
Replace all deadline and event dates String with java.time.LocalDate

felix-ong committed:
Write implementation for Ui, Parser, Storage, TaskList, Command
and *Command classes. Change LocalDate to LocalDateTime.

felix-ong committed:
Integrate previously written classes into Duke

felix-ong committed:
Put all classes in duke package

felix-ong committed:
Add JUnit test for Duke and Parser

felix-ong committed:
Package app as JAR file

felix-ong committed:
Add header comments to *Command classes

felix-ong committed:
Fix code to comply with the given coding standard

felix-ong committed:
Add find feature
Give users a way to find a task by searching for a keyword

felix-ong committed:
Merge branch 'branch-A-JavaDoc'

felix-ong committed:
Merge branch 'branch-A-CodingStandard'

felix-ong committed:
Merge branch 'branch-Level-9'

felix-ong committed:
Merge branch 'add-gradle-support' of https://github.com/felix-ong/ip into branch-A-Gradle

felix-ong committed:
Correct mainClassName in shadowJar

felix-ong committed:
Merge branch 'branch-A-Gradle'

felix-ong committed:
Use CheckStyle and fix coding style violations

felix-ong committed:
Merge branch 'branch-A-CheckStyle'

felix-ong committed:
Add skeletal JavaFX implementation

felix-ong committed:
Add implementation for getResponse in Duke class

felix-ong committed:
Fix checkstyle violations

felix-ong committed:
Package *Command classes and *Task classes and remove Ui class

felix-ong committed:
Add suppressions.xml

felix-ong committed:
Merge branch 'branch-Level-10'

xiangjunn commented on src/main/java/command/AddCommand.java:
```diff
@@ -0,0 +1,46 @@
+        this.task = task;
+    }
+    /**
+     * Executes the specific actions for this command.
```
```suggestion
        this.task = task;
    }
    
    /**
     * Executes the specific actions for this command.
```

xiangjunn commented on src/main/java/command/AddCommand.java:
```diff
@@ -0,0 +1,46 @@
+        this.taskCountMessage = tasks.getTaskCount();
+    }
+
+    /**
+     * Returns false.
+     *
+     * @return false
+     */
+    @Override
+    public boolean isExit() {
+        return false;
+    }
```
Maybe you can provide a description of how this method works? I can't really understand why isExit() should return false. Also maybe you can name it shouldExit() instead of isExit()?

xiangjunn commented on src/main/java/command/DeleteCommand.java:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    /**
+     * Executes the specific actions for this command.
+     *
+     * @param tasks   Handles the list of tasks.
+     * @param storage Handles the saving and loading of tasks.
+     */
```
Perhaps you could stick to the same style for the javadoc comments of the param?
```suggestion
     * Executes the specific actions for this command.
     *
     * @param tasks Handles the list of tasks.
     * @param storage Handles the saving and loading of tasks.
     */
```

xiangjunn commented on src/main/java/controller/MainWindow.java:
```diff
@@ -0,0 +1,52 @@
+import javafx.scene.control.ScrollPane;
+import javafx.scene.control.TextField;
+import javafx.scene.image.Image;
+import javafx.scene.layout.AnchorPane;
+import javafx.scene.layout.VBox;
+/**
+ * Controller for MainWindow. Provides the layout for the other controls.
```
```suggestion
import javafx.scene.layout.AnchorPane;
import javafx.scene.layout.VBox;

/**
 * Controller for MainWindow. Provides the layout for the other controls.
```

xiangjunn commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,54 @@
+    /**
+     * Constructor for Duke with default filepath.
+     */
+    public Duke() {
+        this.storage = new Storage("data/tasks.txt");
+        try {
+            this.tasks = new TaskList(this.storage.loadData());
+        } catch (DukeException e) {
+            this.tasks = new TaskList();
+        }
+    }
+    /**
+     * Constructor for Duke.
+     *
+     * @param filePath Relative path to the storage location.
+     */
+    public Duke(String filePath) {
+        this.storage = new Storage(filePath);
+        try {
+            this.tasks = new TaskList(this.storage.loadData());
+        } catch (DukeException e) {
+            this.tasks = new TaskList();
+        }
+    }
```
There seems to be a duplicate of codes here. Just a recommendation, perhaps you can create a method for this? How about this?
```suggestion
    public Duke() {
        this.setUp("data/tasks.txt);
    }
    
    public Duke(String filePath) {
        this.setUp(filePath);
    }
    
    private void setUp(String filePath) {
        this.storage = new Storage(filePath);
        try {
            this.tasks = new TaskList(this.storage.loadData());
        } catch (DukeException e) {
            this.tasks = new TaskList();
        }
    }
```

xiangjunn commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,94 @@
+     * @return Command to be executed.
+     * @throws DukeException If missing arguments for commands or unknown commands.
+     */
+    public static Command parse(String userInput) throws DukeException {
```
In my opinion, this method seems to be a little too long. If possible, maybe you can try to divide it into more methods.

xiangjunn commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,116 @@
+     * @return A String describing the size of the list of tasks with the appropriate grammar.
+     */
+    public String getTaskCount() {
+        int taskCount = this.tasks.size();
+        return String.format("You have %d %s in the list.%n", taskCount, taskCount > 1 ? "tasks" : "task");
```
Would it be less readable this way? Maybe you can try this?
```suggestion
        int taskCount = this.tasks.size();
        String name = taskCount > 1 ? "tasks" : "task";
        return String.format("You have %d %s in the list.%n", taskCount, name);
```

ganhongyao replied:
I agree with @xiangjunn on this, though maybe a better variable name can be used in place of `name` (I can't think of a better name too, I used "pluralised" for mine 🤣 )

xiangjunn reviewed (commented):
LGTM. Overall, the code is neat and easy to understand. I like how the variable and method names are well-thought of and there is not much ambiguity. Most of the code follow the coding standard. Just some minor issues which are not significant and easy to fix. Good Job! 😄

ganhongyao commented on src/main/java/command/DeleteCommand.java:
```diff
@@ -0,0 +1,46 @@
+     */
+    @Override
+    public void execute(TaskList tasks, Storage storage) throws DukeException {
+        this.task = tasks.deleteTask(storage, this.index);
```
Should we avoid the unnecessary use of `this` as mentioned [here](https://se-education.org/guides/conventions/java/index.html#variables)?

ganhongyao commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,94 @@
+            if (partsLength < 2) {
+                throw new DukeException("☹ OOPS!!! The description of a todo cannot be empty.");
+            }
+            Task todo = new Todo(parts[1]);
```
Is there a reason why you're declaring a `Task` here instead of a `Todo`? Same for the other cases.

ganhongyao commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,94 @@
+            return new ExitCommand();
+        case "todo":
+            if (partsLength < 2) {
+                throw new DukeException("☹ OOPS!!! The description of a todo cannot be empty.");
```
Would it make sense to create a subclass of `DukeException` that is meant for invalid descriptions/arguments (more general)? This would help to reduce the seemingly repeated error messages.

ganhongyao commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,94 @@
+
+    private static String getDescription(String[] parts, int partsLength) {
+        return String.join(" ", Arrays.copyOfRange(parts, 1, partsLength - 3));
+    }
```
Even though this is a private method, would it be good to add in some documentation for this method? Since it may not be obvious what it does at first glance.

ganhongyao commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,116 @@
+    public List<Task> findTask(String keyword) {
+        List<Task> matches = new ArrayList<>();
+        for (Task task : this.tasks) {
+            if (task.getDescription().contains(keyword)) {
```
Not code quality related, but would it be better to do a case-insensitive search?

ganhongyao commented on src/main/java/task/Deadline.java:
```diff
@@ -0,0 +1,56 @@
+     */
+    @Override
+    public String toSaveString() {
+        return "D" + super.toSaveString() + String.format(",%s", this.by);
```
Since you're already using `String.format`, would it be better to just use it for the entire string instead of concatenation?

ganhongyao commented on src/main/java/task/Deadline.java:
```diff
@@ -0,0 +1,56 @@
+     */
+    @Override
+    public String toString() {
+        return "[D]" + super.toString() + " (by: "
+                + this.by.format(DateTimeFormatter.ofPattern("dd MMM yyyy hh:mma")) + ")";
```
Would it be more readable if we used `String.format` here?

ganhongyao reviewed (commented):
Great work! I could not find many things to correct, mostly just minor nits. 👍

felix-ong committed:
Correct mainClassName for shadowJar

felix-ong committed:
Merge branch 'branch-A-Gradle'

felix-ong committed:
Fix styling issues and add ignore case functionality for commands

felix-ong committed:
Add assertions

There are several assumptions in the Storage and Duke classes.
It is assumed that the storage location and the storage field of Duke
has been initialized successfully.

To ensure the assumptions hold, let's add some assertions.

felix-ong committed:
Improve code quality of Parser class

The parse method of Parser was very long, affecting its readability.
Separating it into smaller methods and practicing the
Single Level of Abstraction Principle here will improve the readability.

Let's create new methods to be used in the parse method to make it
more readable and improve the code quality.

felix-ong committed:
Use Streams in TaskList class

Streams can be used in the findTask method to filter out tasks matching
the given keyword and collect them in a list.

This will make the code shorter and more readable.

felix-ong committed:
Merge pull request #1 from felix-ong/branch-A-Assertions

Add assertions

felix-ong committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/duke/Parser.java

felix-ong committed:
Merge pull request #2 from felix-ong/branch-A-CodeQuality

Improve code quality of Parser class

felix-ong committed:
Merge branch 'master' into branch-A-Streams

# Conflicts:
#	src/main/java/duke/TaskList.java

felix-ong committed:
Merge pull request #3 from felix-ong/branch-A-Streams

Use Streams in TaskList class

felix-ong committed:
Add help command

Typing 'help' allows user to see a list of commands and the correct
input format.

felix-ong committed:
Add alias for adding task

't' is an alias for todo, 'd' is an alias for deadline and 'e' is an
alias for event. Having aliases will make the command syntax friendlier.

felix-ong committed:
Merge branch 'branch-BCD-Extension'

felix-ong committed:
Set up Continuous Integration

felix-ong committed:
Change displayPicture to circular and add grey background for user

felix-ong committed:
Merge branch 'branch-A-BetterGUI'

felix-ong committed:
Update README.md with user guide

felix-ong committed:
Merge branch 'branch-A-UserGuide'

felix-ong committed:
Set theme jekyll-theme-minimal

felix-ong committed:
Fix comments

felix-ong committed:
Merge branch 'branch-StyleFixes'

felix-ong committed:
Add javadoc header for Launcher

felix-ong committed:
Merge branch 'branch-StyleFixes'
